api/index.js

```diff
@@ -48,25 +48,32 @@ export default async (req, res) => {
   }
 
   try {
-    const stats = await fetchStats(
+    const statsResp = await fetchStats(
       username,
       parseBoolean(count_private),
       parseBoolean(include_all_commits),
       parseArray(exclude_repo),
     );
+    const stats = statsResp.stats;
 
     const cacheSeconds = clampValue(
       parseInt(cache_seconds || CONSTANTS.FOUR_HOURS, 10),
       CONSTANTS.FOUR_HOURS,
       CONSTANTS.ONE_DAY,
     );
 
-    res.setHeader(
-      "Cache-Control",
-      `max-age=${
-        cacheSeconds / 2
-      }, s-maxage=${cacheSeconds}, stale-while-revalidate=${CONSTANTS.ONE_DAY}`,
-    );
+    if (statsResp.success) {
+      res.setHeader(
+        "Cache-Control",
+        `max-age=${
+          cacheSeconds / 2
+        }, s-maxage=${cacheSeconds}, stale-while-revalidate=${
+          CONSTANTS.ONE_DAY
+        }`,
+      );
+    } else {
+      res.setHeader("Cache-Control", `no-cache, no-store, must-revalidate`); // Don't cache unsuccessful responses.
```

Comment:
Good idea, but I'm just worried that this will make it so that vercel will need to execute more fresh invocations causing increased load when PATs get exhausted. 

For example say we get 10k fresh requests per day, and we cache 70% of the requests, if PATs get exhausted we will render & cache the error card and after 4hour everything will go back to normal. 

But if we do this, all those 10k requests need to be executed by vercel even though they will all respond with error card. 


I think ideal solution would be to decrease the cache seconds on errors: instead of no-cache let's do 1h for error responses (1h since cache will invalidate the same time as PATs get revalidate)

Reply:
That is 100% true I overlooked that fact. Let's close this for now.

Reply:
@anuraghazra In that case we should also change the following code back so that it caches the error:

https://github.com/anuraghazra/github-readme-stats/blob/0ff426d258387ecd0ebbb9206d91b96781e61cf0/api/wakatime.js#L84

https://github.com/anuraghazra/github-readme-stats/blob/0ff426d258387ecd0ebbb9206d91b96781e61cf0/api/top-langs.js#L81

https://github.com/anuraghazra/github-readme-stats/blob/0ff426d258387ecd0ebbb9206d91b96781e61cf0/api/pin.js#L83

https://github.com/anuraghazra/github-readme-stats/blob/0ff426d258387ecd0ebbb9206d91b96781e61cf0/api/index.js#L96

These were merged some time ago and might also be depleting the PATs. However, I suspect this catch block is only triggered when something goes wrong in our code.

Reply:
The fetcher handles errors using `throw` as well as `trycatch`, so that means that errors with the API as well as with our code should trigger the `trycatch` in `./api/index.js`, right?

Trying it out in the browser with a random (nonexistent) username gives a no-cache header, so you're right; this should be changed back, since it has the same effect as whatever the PRs trying to do.

Reply:
@Zo-Bro-23 I changed the code. Please review https://github.com/anuraghazra/github-readme-stats/pull/2448. 👍

Reply:
Can we close this PR now since #2448 does the same thing?

Reply:
This pull request handles errors from the REST API, which will write 0 commits to the card when it fails. While the other pull request handles other errors. I will update this one when #2448 is merged.

Reply:
As far as I understand, API errors will trigger Javascript errors, which means that what the other PR does will apply to API errors too.

https://github.com/anuraghazra/github-readme-stats/blob/4b173001264e9774a2179915519e6129af689625/src/fetchers/stats-fetcher.js#L215-L218

The fetcher is throwing a Javascript error, so GraphQL errors should trigger the `trycatch` block too right? Or am I missing something?